### PR TITLE
Length endpoints

### PR DESCRIPTION
**Pull request by Reettaphant**
Add two new endpoints: 

- One returns word counts of articles grouped into given word count ranges, returns both the count and the range to the client
- Other returns a list of articles without commissioned length (max 500, is this sensible?), and a list with commissioned length sorted according to how much over the commissioned length the word count was

### CODE REVIEW

File: app/controllers/Application.scala [hunk 1]
```diff
@@ -91,4 +91,27 @@ class Application(implicit val wsClient: WSClient, val db: MetricsDB) extends Co
...
   def getNewspaperBookList = APIAuthAction(APIResponse(db.getDistinctNewspaperBooks))
+
+  def getArticlesWithWordCounts() = APIAuthAction { req =>
+    APIResponse {
+      for {
+        articlesWithoutCommissionedLength <- db.getArticlesWithWordCounts(false)(MetricsFilters(req.queryString))
+        articlesWithCommissionedLength <- db.getArticlesWithWordCounts(true)(MetricsFilters(req.queryString))
+      } yield {
+        val resultsOmitted: Boolean = (articlesWithCommissionedLength.length == maxNumberOfArticlesToReturn ||
+          articlesWithoutCommissionedLength.length == maxNumberOfArticlesToReturn)
```
Comment (Reettaphant) [inline app/controllers/Application.scala]: This is not strictly correct, by chance we could just have matched the max number of articles we were to return but I don't think this is a big problem and doing it this way keeps the code neater.

File: app/database/MetricsDB.scala [hunk 2]
```diff
@@ -56,5 +57,58 @@ class MetricsDB(implicit val db: Database) {
...
+    val lowerBoundsToUpperBounds: Map[Int, Int] = Map(
+      0 -> 349,
+      350 -> 649,
+      650 -> 899
+    )
+
+    awaitWithTransformation(db.run(metricsTable
+      .filter(MetricsFilters.metricFilters)
+      .map( metric =>
+        Case
```
Comment (Reettaphant) [inline app/database/MetricsDB.scala]: I would like to refactor this to generate the If statements from the `lowerBoundsToUpperBounds` map but I think this is something for a separate PR if I get the time!

File: app/database/MetricsDB.scala [hunk 3]
```diff
@@ -56,5 +57,58 @@ class MetricsDB(implicit val db: Database) {
...
+
+  def getArticlesWithWordCounts(withCommissionedLength: Boolean)(implicit filters: MetricsFilters):
+    Either[ProductionMetricsError, List[ArticleWordCountResponse]] = {
+
+    val filterFunction = if (withCommissionedLength)
+      MetricsFilters.withCommissionedWordCountFilters
+    else MetricsFilters.withoutCommissionedWordCountFilters
+
+    awaitWithTransformation(db.run(metricsTable.filter(filterFunction)
+      .take(maxNumberOfArticlesToReturn)
```
Comment (RichieAHB) [inline app/database/MetricsDB.scala]: Part of me wonders whether we should do a `COUNT` somewhere too, which means the results omitted would be ok if it landed on 1000 and we could flag how many results were omitted which might be more useful? Although i don't know how tricky that would be ...

Comment (Reettaphant) [reply app/database/MetricsDB.scala]: If that's something that you think would be useful I can easily just fetch all the results from the database, count them and then not send them all to the client.

File: app/controllers/Application.scala [hunk 4]
```diff
@@ -91,4 +91,27 @@ class Application(implicit val wsClient: WSClient, val db: MetricsDB) extends Co
   }
 
   def getNewspaperBookList = APIAuthAction(APIResponse(db.getDistinctNewspaperBooks))
+
+  def getArticlesWithWordCounts() = APIAuthAction { req =>
+    APIResponse {
+      for {
+        articlesWithoutCommissionedLength <- db.getArticlesWithWordCounts(false)(MetricsFilters(req.queryString))
+        articlesWithCommissionedLength <- db.getArticlesWithWordCounts(true)(MetricsFilters(req.queryString))
```
Comment (susiecoleman) [inline app/controllers/Application.scala]: Perhaps
`db.getArticlesWithWordCounts(withCommissionedLength = false)`
`db.getArticlesWithWordCounts(withCommissionedLength = true)`?
As it explicitly tells the reader what the hardcoded boolean represents?

File: app/controllers/Application.scala [hunk 5]
```diff
@@ -91,4 +91,27 @@ class Application(implicit val wsClient: WSClient, val db: MetricsDB) extends Co
...
+        WordCountAPIResponse(articlesWithoutCommissionedLength, articlesWithCommissionedLength, resultsOmitted)
+      }
+    }
+  }
+
+  def getGroupedWordCounts() = APIAuthAction { req =>
+    APIResponse {
+      for {
+        groupedWordCounts <- db.getGroupedWordCounts(MetricsFilters(req.queryString))
+      } yield {
```
Comment (susiecoleman) [inline app/controllers/Application.scala]: Don't need `{}` round the yield

File: app/database/MetricsDB.scala [hunk 6]
```diff
@@ -56,5 +57,58 @@ class MetricsDB(implicit val db: Database) {
       }
 
   def getDistinctNewspaperBooks: Either[ProductionMetricsError, Seq[Option[String]]] =
-    await(db.run(metricsTable.filter(book => !book.newspaperBook.isEmpty && book.newspaperBook =!= "").map(_.newspaperBook).distinct.result))
+    await(db.run(metricsTable.filter(!_.newspaperBook.isEmpty).map(_.newspaperBook).distinct.result))
```
Comment (susiecoleman) [inline app/database/MetricsDB.scala]: Should include `book.newspaperBook =!= ""` in the filter as we want to filter out empty strings

File: app/database/MetricsDB.scala [hunk 7]
```diff
@@ -56,5 +57,58 @@ class MetricsDB(implicit val db: Database) {
...
+
+  def getGroupedWordCounts(implicit filters: MetricsFilters): Either[ProductionMetricsError, List[GroupedWordCount]] = {
```
Comment (susiecoleman) [inline app/database/MetricsDB.scala]: I think documenting in the README how to update word count bands would be a good idea while we're not able to keep them in the config

File: app/database/MetricsDB.scala [hunk 8]
```diff
@@ -56,5 +57,58 @@ class MetricsDB(implicit val db: Database) {
...
+
+    val lowerBoundsToUpperBounds: Map[Int, Int] = Map(
```
Comment (susiecoleman) [inline app/database/MetricsDB.scala]: I think documenting in the README how to update word count bands would be a good idea while we're not able to keep them in the config

File: app/models/APIResponse.scala [hunk 9]
```diff
@@ -3,9 +3,19 @@ package models
 import io.circe._
 import io.circe.generic.semiauto._
 import io.circe.syntax._
+import models.db.ArticleWordCountResponse
 import play.api.Logger
 import play.api.mvc._
 
+
+case class WordCountAPIResponse(articlesWithoutCommissionedLength: List[ArticleWordCountResponse],
+                                articlesWithWordCountResponse: List[ArticleWordCountResponse],
```
Comment (susiecoleman) [inline app/models/APIResponse.scala]: given `articlesWithoutCommissionedLength` and `articlesWithWordCountResponse` are of the same type I would argue that they should either both have `Response` on the end or both drop `Response`.

File: conf/routes [hunk 10]
```diff
@@ -5,6 +5,8 @@ GET     /api/commissioningDesks           controllers.Application.getCommissioni
 GET     /api/inWorkflow/:inWorkflow       controllers.Application.getWorkflowData(inWorkflow: Boolean)
 GET     /api/fork/*newspaperBook          controllers.Application.getForks(newspaperBook: String)
 GET     /api/newspaperBooks               controllers.Application.getNewspaperBookList
+GET     /api/articlesWithWordCounts       controllers.Application.getArticlesWithWordCounts
+GET     /api/groupedWordCounts            controllers.Application.getGroupedWordCounts
```
Comment (susiecoleman) [inline conf/routes]: I wonder if the endpoint names could be more clear? As `/articlesWithWordCounts` makes me think `articlesWithoutWordCounts` would be another endpoint. What about something like `wordcount/articles` `wordcount/bounded`?

File: app/database/MetricsDB.scala (hunk 7 above)
Comment (Reettaphant) [reply app/database/MetricsDB.scala]: Do you think it's hard to find? It's just a map in the appropriate method and I don't think anyone should change them *too* lightheartedly

File: app/database/MetricsDB.scala (hunk 6 above)
Comment (Reettaphant) [reply app/database/MetricsDB.scala]: Yes, git went a bit mad when I rebased this, sorry

**Pull request merged**